### PR TITLE
Calendar Component

### PR DESCRIPTION
**Pull request by Astro-Alien**

### CODE REVIEW

File: app/dialog/dialog.js [hunk 1]
```diff
@@ -135,8 +135,8 @@ export default class Dialog extends crsbinding.classes.ViewBase {
             anchor: "left",
             target: event.target,
             auto_close: true,
-            min_width: "20rem",
-            min_height: "21rem"
+            min_width: "21.88rem",
+            min_height: "24.3rem"
```
Comment (Ashton-Pragma) [inline app/dialog/dialog.js:139]: Make sure these do not end up being half pixels

File: app/dialog/dialog.js [hunk 2]
```diff
@@ -135,8 +135,8 @@ export default class Dialog extends crsbinding.classes.ViewBase {
             anchor: "left",
             target: event.target,
             auto_close: true,
-            min_width: "20rem",
-            min_height: "21rem"
+            min_width: "21.88rem",
```
Comment (Ashton-Pragma) [inline app/dialog/dialog.js]: Make sure these do not end up being half pixels

File: components/calendar/calendar.html [hunk 3]
```diff
@@ -17,19 +17,20 @@
...
         user-select: none;
         height: 100%;
+        grid-row-gap: 0.75rem;
     }
 
-    :focus {
-        outline-color: var(--blue);
+    :focus{
+        outline: 0;
+        box-shadow: inset 0 0 0 2px var(--blue);
```
Comment (Ashton-Pragma) [inline components/calendar/calendar.html]: rem

File: components/calendar/calendar.html [hunk 4]
```diff
@@ -38,11 +39,25 @@
...
     #calendar-toolbar #previous-month, #next-month {
+        width: var(--size);
         cursor: pointer;
+        border-radius: var(--radius);
         border: none;
+        font-family: var(--icon-font-family);
+        font-size: 1.5rem;
+    }
+
+    #calendar-toolbar #previous-month{
```
Comment (Ashton-Pragma) [inline components/calendar/calendar.html]: combine with below

File: components/calendar/calendar.html [hunk 5]
```diff
@@ -52,9 +67,10 @@
         cursor: pointer;
         border-radius: var(--radius);
         border: none;
+        font-family: 'SourceSansPro', sans-serif;
```
Comment (Ashton-Pragma) [inline components/calendar/calendar.html]: why is this necessary

File: components/calendar/calendar.html [hunk 6]
```diff
@@ -91,334 +107,53 @@
         rotate: 90deg;
     }
 
+    #perspectiveElement{
+        height: 19.125rem;
+    }
+
+    @media screen and (max-width: 480px) {
+        #calendar-toolbar  #next-month {
```
Comment (Ashton-Pragma) [inline components/calendar/calendar.html]: combine

File: components/calendar/calendar.html [hunk 7]
```diff
@@ -91,334 +107,53 @@
...
+        }
+    }
+
 </style>
 
 <svg style="display: none" viewBox="0 0 24 24">
     <symbol id="chevron-up" width="24" height="24">
-        <path d="M7.14541 15.6446L12.0063 10.9309L16.8671 15.6446C17.3557 16.1185 18.145 16.1185 18.6336 15.6446C19.1221 15.1708 19.1221 14.4055 18.6336 13.9317L12.8832 8.35535C12.3946 7.88155 11.6054 7.88155 11.1168 8.35535L5.36644 13.9317C4.87785 14.4055 4.87785 15.1708 5.36644 15.6446C5.85503 16.1063 6.65682 16.1185 7.14541 15.6446Z"
-              fill="#272C37"/>
+        <path d="M7.14541 15.6446L12.0063 10.9309L16.8671 15.6446C17.3557 16.1185 18.145 16.1185 18.6336 15.6446C19.1221 15.1708 19.1221 14.4055 18.6336 13.9317L12.8832 8.35535C12.3946 7.88155 11.6054 7.88155 11.1168 8.35535L5.36644 13.9317C4.87785 14.4055 4.87785 15.1708 5.36644 15.6446C5.85503 16.1063 6.65682 16.1185 7.14541 15.6446Z" fill="#272C37"/>
```
Comment (Ashton-Pragma) [inline components/calendar/calendar.html]: Why couldnt we use icon fonts?

File: components/calendar/calendar.js [hunk 8]
```diff
@@ -54,49 +39,66 @@ export default class Calendar extends crsbinding.classes.BindableElement {
...
         await crsbinding.inflationManager.register("calendar-cell", tplCell);
 
         const tplYears = this.shadowRoot.querySelector("#tplYears");
         await crsbinding.inflationManager.register("calendar-years", tplYears);
 
-        this.#tabHandler = this.#tabNavigation.bind(this);
-        this.shadowRoot.addEventListener('keydown', this.#tabHandler);
+        this.#keyboardInputManager = new CalendarKeyboardInputManager(this);
+        this.#changeMonthHandler = this.#changeMonth.bind(this);
+        this.addEventListener("change-month", this.#changeMonthHandler);
```
Comment (Ashton-Pragma) [inline components/calendar/calendar.js:54]: Please use registerEvent instead it auto cleans up on bindable element dispose

File: components/calendar/calendar.js [hunk 9]
```diff
@@ -54,49 +39,66 @@ export default class Calendar extends crsbinding.classes.BindableElement {
...
     async preLoad() {
         this.setProperty("selectedView", "default");
-        this.setProperty("tabindex", "-1");
         await this.#setMonthProperty();
         await this.#setYearProperty();
     }
 
     async attributeChangedCallback(name, oldValue, newValue) {
-        if (newValue == null) return;
+        if (!newValue) return;
```
Comment (Ashton-Pragma) [inline components/calendar/calendar.js]: rather use == null

File: components/dialog/dialog.html [hunk 10]
```diff
@@ -70,6 +70,10 @@
         border-bottom: var(--border);
     }
 
+    #btnToday:hover, #btnDatePickerClose:hover {
+        background-color: var(--grey-s3) !important;
```
Comment (Ashton-Pragma) [inline components/dialog/dialog.html:74]: Be careful of !important its use should only be in very specific scenarios prefer specificity
If your inside a shadow root its a bit safer too

**Review by Ashton-Pragma — CHANGES_REQUESTED** [review]

File: components/calendar/calendar.html (hunk 3 above)
Comment (Astro-Alien) [reply components/calendar/calendar.html]: I converted the 2px to a rem value equivalent to 2px

File: app/dialog/dialog.js (hunk 2 above)
Comment (Astro-Alien) [reply app/dialog/dialog.js]: I updated the min_width: 22rem and min_height:25rem; to prevent it.

File: components/calendar/calendar.html (hunk 5 above)
Comment (Astro-Alien) [reply components/calendar/calendar.html]: the month and year font-family isn't being set by default which leads to the font being Arial instead of 'SourceSansPro', sans-serif which causes the fonts to look different.

File: components/calendar/calendar.html (hunk 4 above)
Comment (Astro-Alien) [reply components/calendar/calendar.html]: I combined the styles but changed the query to #calendar-toolbar #previous-month, #year to gain the same result.

File: components/calendar/calendar.html (hunk 6 above)
Comment (Astro-Alien) [reply components/calendar/calendar.html]: I combined the styles but changed the query to #calendar-toolbar #previous-month, #year to gain the same result.

File: components/calendar/calendar.html (hunk 7 above)
Comment (Astro-Alien) [reply components/calendar/calendar.html]: I removed the unused svg from the html script.

File: components/calendar/calendar.js (hunk 8 above)
Comment (Astro-Alien) [reply components/calendar/calendar.js:54]: I attempted to use registerEvent and tested it, it seemed to cause memory leaks as it wasn't being cleaned by the bindable elements dispose, it could be something similar to what was happening with the bindable events.

File: components/calendar/calendar.js (hunk 9 above)
Comment (Astro-Alien) [reply components/calendar/calendar.js]: I replaced the if(!newValue) return with if( newValue == null) return

**Review by Ashton-Pragma — APPROVED** [review]
